### ISSUE TITLE
Parser error with `:lifetime` param in expanded code

### ISSUE THREAD

**Issue by danielhenrymantilla** [dtolnay/paste#26]
Minimal repro:

```rust
macro_rules! m {(
    $lt:lifetime
) => (
    ::paste::item! {
        impl<$lt> () {}
    }
)}

m! { 'lt }
```

leads to a parser error saying that:

  - `1.45.0-nightly (90931d9b3 2020-04-28)`
    ```text
    error: lifetime in trait object type must be followed by `+`
    ```

  - `1.43.0 (3532cf738 2020-03-17)`
    ```text
    error: expected type, found `'lt`
    ```

  - A run with `cargo bisect` seems to indicate this bug has always been present.
___

EDIT: After some testing, here is a minimal proc-macro repro (that is, if we replace `::paste::item!` by `foo!`, the same parsing error is triggered):

```rust
#[proc_macro] pub
fn foo (it: TokenStream) -> TokenStream
{
    it  .into_iter()
        .map(|tt| tt) // without this line, there is no error (I imagine due to some specialization) 
        .collect::<TokenStream>()
      //.to_string().parse().unwrap() /* workaround! */
}
```

This minimal repro suggests that this is not really an issue with `paste`, but with some proc-macro `:lifetime` capture interaction... Since you know the internals of a `TokenStream` better, @dtolnay, I'll let you investigate it further and post an appropriate issue on rust-lang/rust.

**Comment by danielhenrymantilla** [comment]
A super-hacky workaround when expanding to an `item` and when not using `$crate` is to use `.to_string().parse().unwrap()` on the `.collect()`-ed `TokenStream`, so as to construct a `TokenStream` that `rustc` likes.

**Comment by danielhenrymantilla** [comment]
For those needing a workaround, "flattening" / unwrapping the `Group`ed `:lifetime` into the two tokens that compose it seems to solve the issue. That is, wrapping the returned `TokenStream` within the following function avoids the compilation error:

```rust


fn fix_lifetime_capture_hack (ts: TokenStream)
  -> TokenStream
{
    use TokenTree as TT;
    ts.into_iter().map(|tt| match tt {
        | TT::Group(group) if true
            && group.delimiter() == Delimiter::None
            && matches!(
                group.stream().into_iter().next(),
                Some(TT::Punct(ref punct)) if punct.as_char() == '\''
            )
        => group.stream(),

        | TT::Group(ref group) => {
            let span = tt.span();
            let mut tt = TT::Group(Group::new(
                group.delimiter(),
                fix_lifetime_capture_hack(group.stream()),
            ));
            tt.set_span(span);
            iter::once(tt).collect()
        },

        | _ => iter::once(tt).collect(),
    }).collect()
}
```

**Comment by dtolnay** [comment]
This seems like a compiler bug, but I've released 0.1.12 with a workaround.
